.gitlab-ci.yml

```diff
@@ -21,137 +77,139 @@ services:
     alias: quilc
     command: ["-R"]
 
+
 ####################################################################################################
 # EVERY-COMMIT JOBS
 ####################################################################################################
 
-# build the Sphinx documentation
-# for all Python jobs, see rigetti/gitlab-pipelines/python.gitlab-ci.yml
-build-docs:
-  extends: .python
+Test Docs:
+  stage: test
   script:
+    - *install-dependencies
     - apt-get update && apt-get install -y pandoc
-    - make docs
-
-# validate code format against Black
-check-format:
-  extends: .python
-  script:
-    - make check-format
-
-# validate code style against flake8
-check-style:
-  extends: .python
-  script:
-    - make check-style
+    - poetry run make docs
 
-# perform static type checking with mypy
-check-types:
-  extends: .python
+Style:
+  stage: test
   script:
-    - make check-types
-
-# run the unit tests with Python 3.7, and report coverage
-test-py37:
-  extends: .python
+    - *install-dependencies
+    - poetry run make check-all
+  rules:
+    # Skip this job if it was triggered by a tag
+    - if: "$CI_COMMIT_TAG"
+      when: never
+    - when: always
+
+Test (3.7):
+  stage: test
+  image: python:3.7
+  coverage: '/TOTAL.*?(\d+)\%/'
   script:
-    - make test
-  coverage: '/TOTAL.*\s(\d+)%/'
+    - *install-dependencies
+    - poetry run make test
 
-# run the unit tests with Python 3.8
-test-py38:
-  extends: .python
+Test (3.8):
+  stage: test
   image: python:3.8
+  coverage: '/TOTAL.*?(\d+)\%/'
   script:
-    - make test
+    - *install-dependencies
+    - poetry run make test
 
-# test for coverage using coveralls
-test-coverage:
-  extends: .python
+Coverage:
+  stage: test
   script:
-    - make coverage
+    - *install-dependencies
+    - poetry run make coverage
   allow_failure: true
 
+
 ####################################################################################################
 # MASTER-ONLY JOBS
 ####################################################################################################
 
-# create a Docker image from this master commit (see rigetti/gitlab-pipelines/docker.gitlab-ci.yml)
-docker-edge:
-  extends: .docker-edge
-
-# upload the source distribution of this master commit to TestPyPI
-upload-testpypi:
-  extends: .python
+PyPI Publish (Edge):
```

Comment:
Note the pattern: there's a "PyPI Publish" followed by a "Docker Publish" for the following scenarios:
- branches (manual)
- commits to master (automatic)
- releases (manual)